File: interfaces/pkcs7.js
/* pkcs7.js
 *
 * Copyright (C) 2006-2024 wolfSSL Inc.
 *
 * This file is part of wolfSSL.
 *
 * wolfSSL is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * wolfSSL is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1335, USA
 */
const wolfcrypt = require( '../build/Release/wolfcrypt' )

class WolfSSL_PKCS7
{
  /**
   * Creates a new PKCS7 structure by calling sizeof_PKCS7 and wc_PKCS7_Init
   *
   * @remarks free must be called to free the ecc PKCS7 data
   */
  constructor()
  {
    this.pkcs7 = Buffer.alloc( wolfcrypt.sizeof_PKCS7() )

    let ret = wolfcrypt.wc_PKCS7_Init( this.pkcs7 )

    if ( ret != 0 )
    {
      throw `Failed to wc_PKCS7_Init ${ ret }`
    }
  }

  /**
   * Adds a certificate to the PKCS7 structure
   *
   * @param cert The cert to add.
   *
   * @throws {Error} If the PKCS7 structure is not allocated.
   *
   * @throws {Error} If the cert is not a Buffer.
   *
   * @throws {Error} If wc_PKCS7_AddCertificate fails.
   */
  AddCertificate( cert )
  {
    if ( this.pkcs7 == null )
    {
      throw 'Pkcs7 not allocated'
    }

    if ( !Buffer.isBuffer( cert ) )
    {
      throw `cert must be a Buffer`
    }

    let ret = wolfcrypt.wc_PKCS7_AddCertificate( this.pkcs7, cert, cert.length )

    if ( ret != 0 )
    {
      throw `Failed to wc_PKCS7_AddCertificate ${ ret }`
    }
  }

  /**
   * Encodes the provided data using the provided key
   *
   * @param data The data to encode.
   *
   * @param key The key to use, as a Buffer.
   *
   * @returns The encoded data Buffer.
   *
   * @throws {Error} If the PKCS7 structure is not allocated.
   *
   * @throws {Error} If the data is not a string or Buffer.
   *
   * @throws {Error} If the key is not a Buffer.
   *
   * @throws {Error} If wc_PKCS7_EncodeData fails.
   */
  EncodeData( data, key )
  {
    if ( this.pkcs7 == null )
    {
      throw 'Pkcs7 not allocated'
    }

    if ( typeof data == 'string' )
    {
      data = Buffer.from( data )
    }

    if ( !Buffer.isBuffer( data ) )
    {
      throw `data must be a string or Buffer`
    }

    if ( !Buffer.isBuffer( key ) )
    {
      throw `cert must be a Buffer`
    }

    let outBuf = Buffer.alloc( 4000 )

    let ret = wolfcrypt.wc_PKCS7_EncodeData( this.pkcs7, data, data.length, key, key.length, outBuf, outBuf.length )

    if ( ret <= 0 )
    {
      throw `Failed to wc_PKCS7_EncodeData ${ ret }`
    }

    outBuf = outBuf.subarray( 0, ret )

    return outBuf
  }

  /**
   * Encodes and signs the provided data using the provided key, keySum and hashSum
   *
   * @param data The data to encode and sign.
   *
   * @param key The key to use, as a Buffer.
   *
   * @param keySum The key algorithm to use, RSA, ECDSA, ED25519, etc.
   *
   * @param keySum The hash algorithm to use, MD5, SHA, SHAKE256, etc.
   *
   * @returns The signed data Buffer.
   *
   * @throws {Error} If the PKCS7 structure is not allocated.
   *
   * @throws {Error} If the data is not a string or Buffer.
   *
   * @throws {Error} If the key is not a Buffer.
   *
   * @throws {Error} If keySum or hashSum are invalid.
   *
   * @throws {Error} If wc_PKCS7_EncodeSignedData fails.
   */
  EncodeSignedData( data, key, keySum, hashSum )
  {
    if ( this.pkcs7 == null )
    {
      throw 'Pkcs7 not allocated'
    }

    if ( typeof data == 'string' )
    {
      data = Buffer.from( data )
    }

    if ( !Buffer.isBuffer( data ) )
    {
      throw `data must be a string or Buffer`
    }

    if ( !Buffer.isBuffer( key ) )
    {
      throw `cert must be a Buffer`
    }

    let keySumType = wolfcrypt.typeof_Key_Sum( keySum )

    if ( keySumType < 0 )
    {
      throw `Invalid keySum`
    }

    let hashSumType = wolfcrypt.typeof_Hash_Sum( hashSum )

    if ( hashSumType < 0 )
    {
      throw `Invalid hashSum`
    }

    let outBuf = Buffer.alloc( 4000 )

    let ret = wolfcrypt.wc_PKCS7_EncodeSignedData( this.pkcs7, data, data.length, key, key.length, keySumType, hashSumType, outBuf, outBuf.length )

    if ( ret <= 0 )
    {
      throw `Failed to wc_PKCS7_EncodeSignedData ${ ret }`
    }

    outBuf = outBuf.subarray( 0, ret )

    return outBuf
  }

  /**
   * Verifies a signed data Buffer and loads it into the PKCS7 struct
   *
   * @param data The signed data to verify.
   *
   * @throws {Error} If the PKCS7 structure is not allocated.
   *
   * @throws {Error} If the data is not a Buffer.
   *
   * @throws {Error} If wc_PKCS7_VerifySignedData fails.
   */
  VerifySignedData( data )
  {
    if ( this.pkcs7 == null )
    {
      throw 'Pkcs7 not allocated'
    }

    if ( !Buffer.isBuffer( data ) )
    {
      throw `data must be a Buffer`
    }

    let ret = wolfcrypt.wc_PKCS7_VerifySignedData( this.pkcs7, data, data.length )

    if ( ret != 0 )
    {
      throw `Failed to wc_PKCS7_VerifySignedData ${ ret }`
    }
  }

  /**
   * Retreives an attribute from the PKCS7 structure by its oid
   *
   * @param oid The oid of the desired attribute.
   *
   * @returns The attribute value as a Buffer.
   *
   * @throws {Error} If the PKCS7 structure is not allocated.
   *
   * @throws {Error} If the oid is not a string or Buffer.
   *
   * @throws {Error} If wc_PKCS7_GetAttributeValue fails.
   */
  GetAttributeValue( oid )
  {
    if ( this.pkcs7 == null )
    {
      throw 'Pkcs7 not allocated'
    }

    if ( typeof oid == 'string' )
    {
      oid = Buffer.from( oid )
    }

    if ( !Buffer.isBuffer( oid ) )
    {
      throw `oid must be a Buffer or string`
    }

    let outBuf = Buffer.alloc( wolfcrypt.sizeof_wc_PKCS7_GetAttributeValue( this.pkcs7, oid, oid.length ) )

    let ret = wolfcrypt.wc_PKCS7_GetAttributeValue(  this.pkcs7, oid, oid.length, outBuf, outBuf.length )

    if ( ret <= 0 )
    {
      throw `Failed to wc_PKCS7_GetAttributeValue ${ ret }`
    }

    return outBuf
  }

  /**
   * Retreives the SID from the PKCS7 structure
   *
   * @returns The SID value as a Buffer.
   *
   * @throws {Error} If the PKCS7 structure is not allocated.
   *
   * @throws {Error} If wc_PKCS7_GetSignerSID fails.
   */
  GetSignerSID()
  {
    if ( this.pkcs7 == null )
    {
      throw 'Pkcs7 not allocated'
    }

    let outBuf = Buffer.alloc( wolfcrypt.sizeof_wc_PKCS7_GetSignerSID( this.pkcs7 ) )

    let ret = wolfcrypt.wc_PKCS7_GetSignerSID( this.pkcs7, outBuf, outBuf.length )

    if ( ret <= 0 )
    {
      throw `Failed to wc_PKCS7_GetSignerSID ${ ret }`
    }

    return outBuf
  }

  /**
   * Frees the data allocated by the PKCS7 structure
   *
   * @throws {Error} If PKCS7 structure is not allocated.
   *
   * @throws {Error} If wc_ecc_free fails.
   */
  free()
  {
    if ( this.pkcs7 == null )
    {
      throw 'Pkcs7 not allocated'
    }

    wolfcrypt.wc_PKCS7_Free( this.pkcs7 )

    this.pkcs7 = null
  }
}

exports.WolfSSL_PKCS7 = WolfSSL_PKCS7

File: addon/wolfcrypt/pkcs7.cpp
/* pkcs7.cpp
 *
 * Copyright (C) 2006-2022 wolfSSL Inc.
 *
 * This file is part of wolfSSL.
 *
 * wolfSSL is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * wolfSSL is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1335, USA
 */
#include "./h/pkcs7.h"

Napi::Number sizeof_PKCS7(const Napi::CallbackInfo& info)
{
  Napi::Env env = info.Env();

  return Napi::Number::New( env, sizeof( PKCS7 ) );
}

Napi::Number typeof_Key_Sum(const Napi::CallbackInfo& info)
{
  int ret = -1;
  Napi::Env env = info.Env();
  std::string type = info[0].As<Napi::String>().Utf8Value();

  if ( strcmp( type.c_str(), "DSA" ) == 0 )
  {
    ret = DSAk;
  }
  else if ( strcmp( type.c_str(), "RSA" ) == 0 )
  {
    ret = RSAk;
  }
  else if ( strcmp( type.c_str(), "ECDSA" ) == 0 )
  {
    ret = ECDSAk;
  }
  else if ( strcmp( type.c_str(), "ED25519" ) == 0 )
  {
    ret = ED25519k;
  }
  else if ( strcmp( type.c_str(), "X25519" ) == 0 )
  {
    ret = X25519k;
  }
  else if ( strcmp( type.c_str(), "ED448" ) == 0 )
  {
    ret = ED448k;
  }
  else if ( strcmp( type.c_str(), "X448" ) == 0 )
  {
    ret = X448k;
  }
  else if ( strcmp( type.c_str(), "DH" ) == 0 )
  {
    ret = DHk;
  }

  return Napi::Number::New( env, ret );
}

Napi::Number typeof_Hash_Sum(const Napi::CallbackInfo& info)
{
  int ret = -1;
  Napi::Env env = info.Env();
  std::string type = info[0].As<Napi::String>().Utf8Value();

  if ( strcmp( type.c_str(), "MD2" ) == 0 )
  {
    ret = MD2h;
  }
  else if ( strcmp( type.c_str(), "MD5" ) == 0 )
  {
    ret = MD5h;
  }
  else if ( strcmp( type.c_str(), "SHA" ) == 0 )
  {
    ret = SHAh;
  }
  else if ( strcmp( type.c_str(), "SHA224" ) == 0 )
  {
    ret = SHA224h;
  }
  else if ( strcmp( type.c_str(), "SHA256" ) == 0 )
  {
    ret = SHA256h;
  }
  else if ( strcmp( type.c_str(), "SHA384" ) == 0 )
  {
    ret = SHA384h;
  }
  else if ( strcmp( type.c_str(), "SHA512" ) == 0 )
  {
    ret = SHA512h;
  }
  else if ( strcmp( type.c_str(), "SHA512_224" ) == 0 )
  {
    ret = SHA512_224h;
  }
  else if ( strcmp( type.c_str(), "SHA512_256" ) == 0 )
  {
    ret = SHA512_256h;
  }
  else if ( strcmp( type.c_str(), "SHA3_224" ) == 0 )
  {
    ret = SHA3_224h;
  }
  else if ( strcmp( type.c_str(), "SHA3_256" ) == 0 )
  {
    ret = SHA3_256h;
  }
  else if ( strcmp( type.c_str(), "SHA3_384" ) == 0 )
  {
    ret = SHA3_384h;
  }
  else if ( strcmp( type.c_str(), "SHA3_512" ) == 0 )
  {
    ret = SHA3_512h;
  }
  else if ( strcmp( type.c_str(), "SHAKE128" ) == 0 )
  {
    ret = SHAKE128h;
  }
  else if ( strcmp( type.c_str(), "SHAKE256" ) == 0 )
  {
    ret = SHAKE256h;
  }

  return Napi::Number::New( env, ret );
}

Napi::Number bind_wc_PKCS7_Init(const Napi::CallbackInfo& info)
{
  int ret;
  Napi::Env env = info.Env();
  PKCS7* pkcs7 = (PKCS7*)( info[0].As<Napi::Uint8Array>().Data() );

  ret = wc_PKCS7_Init( pkcs7, NULL, INVALID_DEVID );

  return Napi::Number::New( env, ret );
}

Napi::Number bind_wc_PKCS7_InitWithCert(const Napi::CallbackInfo& info)
{
  int ret;
  Napi::Env env = info.Env();
  PKCS7* pkcs7 = (PKCS7*)( info[0].As<Napi::Uint8Array>().Data() );
  uint8_t* cert = info[1].As<Napi::Uint8Array>().Data();
  int cert_size = info[2].As<Napi::Number>().Int32Value();

  ret = wc_PKCS7_InitWithCert( pkcs7, cert, cert_size );

  return Napi::Number::New( env, ret );
}

Napi::Number bind_wc_PKCS7_AddCertificate(const Napi::CallbackInfo& info)
{
  int ret;
  Napi::Env env = info.Env();
  PKCS7* pkcs7 = (PKCS7*)( info[0].As<Napi::Uint8Array>().Data() );
  uint8_t* cert = info[1].As<Napi::Uint8Array>().Data();
  int cert_size = info[2].As<Napi::Number>().Int32Value();

  ret = wc_PKCS7_AddCertificate( pkcs7, cert, cert_size );

  return Napi::Number::New( env, ret );
}

Napi::Number bind_wc_PKCS7_EncodeData(const Napi::CallbackInfo& info)
{
  int ret;
  Napi::Env env = info.Env();
  PKCS7* pkcs7 = (PKCS7*)( info[0].As<Napi::Uint8Array>().Data() );
  uint8_t* data = info[1].As<Napi::Uint8Array>().Data();
  int data_size = info[2].As<Napi::Number>().Int32Value();
  uint8_t* key = info[3].As<Napi::Uint8Array>().Data();
  int key_size = info[4].As<Napi::Number>().Int32Value();
  uint8_t* output = info[5].As<Napi::Uint8Array>().Data();
  int output_size = info[6].As<Napi::Number>().Int32Value();

  pkcs7->content = data;
  pkcs7->contentSz = data_size;
  pkcs7->privateKey = key;
  pkcs7->privateKeySz = key_size;

  ret = wc_PKCS7_EncodeData( pkcs7, output, output_size );

  return Napi::Number::New( env, ret );
}

Napi::Number bind_wc_PKCS7_EncodeSignedData(const Napi::CallbackInfo& info)
{
  int ret;
  Napi::Env env = info.Env();
  WC_RNG rng;
  PKCS7* pkcs7 = (PKCS7*)( info[0].As<Napi::Uint8Array>().Data() );
  uint8_t* data = info[1].As<Napi::Uint8Array>().Data();
  int data_size = info[2].As<Napi::Number>().Int32Value();
  uint8_t* key = info[3].As<Napi::Uint8Array>().Data();
  int key_size = info[4].As<Napi::Number>().Int32Value();
  int key_sum = info[5].As<Napi::Number>().Int32Value();
  int hash_sum = info[6].As<Napi::Number>().Int32Value();
  uint8_t* output = info[7].As<Napi::Uint8Array>().Data();
  int output_size = info[8].As<Napi::Number>().Int32Value();

  wc_InitRng( &rng );

  ret = wc_PKCS7_SetSignerIdentifierType( pkcs7, CMS_SKID );

  if ( ret != 0 )
  {
    return Napi::Number::New( env, ret );
  }

  pkcs7->content = data;
  pkcs7->contentSz = data_size;
  pkcs7->privateKey = key;
  pkcs7->privateKeySz = key_size;
  pkcs7->encryptOID = key_sum;
  pkcs7->publicKeyOID = key_sum;
  pkcs7->hashOID = hash_sum;
  pkcs7->rng = &rng;

  ret = wc_PKCS7_EncodeSignedData( pkcs7, output, output_size );

  wc_FreeRng( &rng );

  return Napi::Number::New( env, ret );
}

Napi::Number bind_wc_PKCS7_VerifySignedData(const Napi::CallbackInfo& info)
{
  int ret;
  Napi::Env env = info.Env();
  PKCS7* pkcs7 = (PKCS7*)( info[0].As<Napi::Uint8Array>().Data() );
  uint8_t* in = info[1].As<Napi::Uint8Array>().Data();
  int in_size = info[2].As<Napi::Number>().Int32Value();

  ret = wc_PKCS7_VerifySignedData( pkcs7, in, in_size );

  return Napi::Number::New( env, ret );
}

Napi::Number sizeof_wc_PKCS7_GetAttributeValue(const Napi::CallbackInfo& info)
{
  int ret;
  Napi::Env env = info.Env();
  PKCS7* pkcs7 = (PKCS7*)( info[0].As<Napi::Uint8Array>().Data() );
  uint8_t* oid = info[1].As<Napi::Uint8Array>().Data();
  unsigned int oid_size = info[2].As<Napi::Number>().Int32Value();
  unsigned int out_size;

  ret = wc_PKCS7_GetAttributeValue( pkcs7, oid, oid_size, NULL, &out_size );

  if ( ret != LENGTH_ONLY_E )
  {
    return Napi::Number::New( env, ret );
  }

  return Napi::Number::New( env, out_size );
}

Napi::Number bind_wc_PKCS7_GetAttributeValue(const Napi::CallbackInfo& info)
{
  int ret;
  Napi::Env env = info.Env();
  PKCS7* pkcs7 = (PKCS7*)( info[0].As<Napi::Uint8Array>().Data() );
  uint8_t* oid = info[1].As<Napi::Uint8Array>().Data();
  unsigned int oid_size = info[2].As<Napi::Number>().Int32Value();
  uint8_t* out = info[3].As<Napi::Uint8Array>().Data();
  unsigned int out_size = info[4].As<Napi::Number>().Int32Value();

  ret = wc_PKCS7_GetAttributeValue( pkcs7, oid, oid_size, out, &out_size );

  if ( ret < 0 )
  {
    return Napi::Number::New( env, ret );
  }

  return Napi::Number::New( env, out_size );
}

Napi::Number sizeof_wc_PKCS7_GetSignerSID(const Napi::CallbackInfo& info)
{
  int ret;
  Napi::Env env = info.Env();
  PKCS7* pkcs7 = (PKCS7*)( info[0].As<Napi::Uint8Array>().Data() );
  unsigned int out_size;

  ret = wc_PKCS7_GetSignerSID( pkcs7, NULL, &out_size );

  if ( ret != LENGTH_ONLY_E )
  {
    return Napi::Number::New( env, ret );
  }

  return Napi::Number::New( env, out_size );
}

Napi::Number bind_wc_PKCS7_GetSignerSID(const Napi::CallbackInfo& info)
{
  int ret;
  Napi::Env env = info.Env();
  PKCS7* pkcs7 = (PKCS7*)( info[0].As<Napi::Uint8Array>().Data() );
  uint8_t* out = info[1].As<Napi::Uint8Array>().Data();
  unsigned int out_size = info[2].As<Napi::Number>().Int32Value();

  ret = wc_PKCS7_GetSignerSID( pkcs7, out, &out_size );

  if ( ret < 0 )
  {
    return Napi::Number::New( env, ret );
  }

  return Napi::Number::New( env, out_size );
}

void bind_wc_PKCS7_Free(const Napi::CallbackInfo& info)
{
  PKCS7* pkcs7 = (PKCS7*)( info[0].As<Napi::Uint8Array>().Data() );

  wc_PKCS7_Free( pkcs7 );
}
